projects/rocprofiler-sdk/tests/bin/attachment-test/attachment_test.cpp

```diff
@@ -109,46 +105,104 @@ main(int /*argc*/, char** /*argv*/)
+                      << " on device " << device_id << "...\n";
             roctxRangePop();  // Removed - ROCTx not linked
             break;
         }
 
         roctxRangePop();  // Removed - ROCTx not linked
 
-        std::cout << "Iteration " << (iter + 1) << "/" << num_iterations << " completed"
-                  << std::endl;
-
         // Small delay between iterations
         std::this_thread::sleep_for(std::chrono::milliseconds(500));
     }
 
-    std::cout << "Kernel execution loop completed" << std::endl;
+    std::cout << "Kernel execution loop completed for thread " << tid << " with stream "
+              << stream_id << " on device " << device_id << "...\n";
 
     // Cleanup
-    err = hipFree(d_data);
-    if(err != hipSuccess)
+    HIP_ASSERT(hipFree(d_data));
+    delete[] h_data;
+}
+
+int
+main(int argc, char** argv)
+{
+    size_t nthreads{32};
+    size_t nstreams{8};
+    int    ndevices{0};
+    for(int i = 1; i < argc; ++i)
     {
-        std::cerr << "Warning: Failed to free device memory" << std::endl;
+        auto _arg = std::string{argv[i]};
+        if(_arg == "?" || _arg == "-h" || _arg == "--help")
+        {
+            fprintf(stderr,
+                    "usage: attachment-test [NUM_THREADS (%zu)] [NUM_STREAMS (%zu)] "
+                    "[NUM_DEVICES (%d)]\n",
+                    nthreads,
+                    nstreams,
+                    ndevices);
+            exit(EXIT_SUCCESS);
+        }
     }
-    delete[] h_data;
+    if(argc > 1) nthreads = std::atoll(argv[1]);
+    if(argc > 2) nstreams = std::atoll(argv[2]);
+    if(argc > 3) ndevices = std::stoi(argv[3]);
+
+    std::cout << "Attachment test app started with PID: " << getpid() << std::endl;
+
+    // Initialize HIP
+    int device_count = 0;
+    HIP_ASSERT(hipGetDeviceCount(&device_count));
+    if(device_count == 0)
+    {
+        std::cerr << "No HIP devices found or error getting device count" << std::endl;
+        return 1;
+    }
+    // Default ndecives to device_count. Ensure that we do not use more devices than are available
+    ndevices = ndevices == 0 ? device_count : ndevices;
+    if(ndevices > device_count)
+    {
+        std::cout << "Using " << device_count << " HIP devices instead of the requested "
+                  << ndevices << "\n";
+        ndevices = device_count;
+    }
+
+    std::cout << "After first call " << getpid() << std::endl;
+
+    auto                     _threads = std::vector<std::thread>{};
+    std::vector<hipStream_t> _streams(nstreams);
+    _threads.reserve(nthreads);
+
+    for(size_t i = 0; i < nstreams; ++i)
+        HIP_ASSERT(hipStreamCreate(&_streams[i]));
+    for(size_t i = 0; i < nthreads; ++i)
+        _threads.emplace_back(
+            execute_kernels, i, _streams[i % nstreams], i % nstreams, i % ndevices);
+    for(auto& itr : _threads)
+        itr.join();
+
+    // Destroy streams
+    for(size_t i = 0; i < nstreams; ++i)
```

Comment:
```suggestion
    for(auto itr : _streams)
```
Use range-based for loop whenever possible